### PR TITLE
(#55) Config parsing with once_cell

### PR DESCRIPTION
**Pull request by sigaloid**
A better approach to this than my config branch, utilizing [once-cell](https://lib.rs/once_cell). Once once_cell lands in the standard library, we can use it - for now we don't want a reliance on the nightly toolchain so I pulled in that crate.

Crates added:
* `toml` (config parsing) (TODO: add file-based config to wiki)
* `once_cell` (lazily initializing the immutable static)

Performance changes: None that I can think of. 

This PR creates a config struct that will do this at runtime, once:
* read from environment variable and parse into settings struct (higher priority)
* read from `ferrit.toml` and parse into settings struct (lower priority)
* saves struct as global static immutable - unchanging so we should not have any gremlins relating to global state
* no more future calls to environment variables or reading config - it is stored as-is. This might improve performance and reduce the constant calls to environment variables on every request

### CODE REVIEW

**Comment by Daniel-Valentine** [comment]
This is so much nicer than what I had planned and worked on. We may want to go with this.

**Comment by sigaloid** [comment]
Fixed CI and silly typo as well as reduced code duplication

**Comment by Daniel-Valentine** [comment]
Looks good, although there are some stylistic issues that I'll point out. We don't have a style-guide yet but I'll write one soon. I'll approve and you can merge at your leisure.

File: src/config.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
+use std::env::var;
+
+use once_cell::sync::Lazy;
+
+// Waiting for https://github.com/rust-lang/rust/issues/74465 to land, so we can reduce reliance on once_cell
+static CONFIG: Lazy<Config> = Lazy::new(|| Config::load());
+
+#[derive(Default, serde::Deserialize)]
+pub struct Config {
```
Comment (Daniel-Valentine) [inline src/config.rs:9]: Since this is exported, let's add a `///` for this struct.

File: src/config.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
+use std::env::var;
+
+use once_cell::sync::Lazy;
+
+// Waiting for https://github.com/rust-lang/rust/issues/74465 to land, so we can reduce reliance on once_cell
```
Comment (Daniel-Valentine) [inline src/config.rs:5]: Let's limit the length of comments wherever possible to 80 chars.

File: src/config.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
+use std::env::var;
+
```
Comment (Daniel-Valentine) [inline src/config.rs:2]: Remove the space + alphabetize the imports.

File: src/config.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	default_comment_sort: Option<String>,
+	default_post_sort: Option<String>,
+	default_show_nsfw: Option<String>,
+	default_blur_nsfw: Option<String>,
+	default_use_hls: Option<String>,
+	default_hide_hls_notification: Option<String>,
+}
+
+impl Config {
+	pub fn load() -> Self {
```
Comment (Daniel-Valentine) [inline src/config.rs:24]: Since this is exported, let's comment this with `///`.

File: src/config.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+			default_post_sort: parse("FERRIT_DEFAULT_POST_SORT"),
+			default_wide: parse("FERRIT_DEFAULT_WIDE"),
+			default_comment_sort: parse("FERRIT_DEFAULT_COMMENT_SORT"),
+			default_show_nsfw: parse("FERRIT_DEFAULT_SHOW_NSFW"),
+			default_blur_nsfw: parse("FERRIT_DEFAULT_BLUR_NSFW"),
+			default_use_hls: parse("FERRIT_DEFAULT_USE_HLS"),
+			default_hide_hls_notification: parse("FERRIT_DEFAULT_HIDE_HLS"),
+		}
+	}
+}
```
Comment (Daniel-Valentine) [inline src/config.rs:50]: Add a newline between the last bracket and the function signature.

```suggestion
}

fn get_setting_from_config(name: &str, config: &Config) -> Option<String> {
```

File: src/config.rs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+		"FERRIT_DEFAULT_SHOW_NSFW" => config.default_show_nsfw.clone(),
+		"FERRIT_DEFAULT_BLUR_NSFW" => config.default_blur_nsfw.clone(),
+		"FERRIT_DEFAULT_USE_HLS" => config.default_use_hls.clone(),
+		"FERRIT_DEFAULT_HIDE_HLS_NOTIFICATION" => config.default_hide_hls_notification.clone(),
+		"FERRIT_DEFAULT_WIDE" => config.default_wide.clone(),
+		_ => None,
+	}
+}
+
+/// Retrieve setting from environment variable or config file
```
Comment (Daniel-Valentine) [inline src/config.rs:68]: 1. Comment needs a terminal period.
2. `Retrieve` -> `Retrieves`

**Review by Daniel-Valentine — APPROVED** [review]

File: src/config.rs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+impl Config {
+	pub fn load() -> Self {
+		// Read from ferris.toml config file. If for any reason, it fails, the default `Config` is used (all None values)
+		let config: Config = toml::from_str(&std::fs::read_to_string("ferris.toml").unwrap_or_default()).unwrap_or_default();
+		// This function defines the order of preference - first check for environment variables, then check the config, then if both are `None`,
+		// return a `None` via the `map_or_else` function
+		let parse = |key: &str| -> Option<String> { var(key).ok().map_or_else(|| get_setting_from_config(key, &config), Some) };
+		Self {
+			sfw_only: parse("FERRIT_SFW_ONLY"),
+			default_theme: parse("FERRIT_DEFAULT_THEME"),
```
[the remark this answers is not in the document]
Comment (sigaloid) [reply src/config.rs:45]: It does, here: https://github.com/ferritreader/ferrit/pull/59/files#diff-cba64c21ab992eaad29fce147a08f4560a4769bc14682b8a96081a5fd02dbecdR40

**Comment by sigaloid** [comment]
Once this is merged I'll write some tests for it with [sealed_test](https://lib.rs/crates/sealed_test)

**Review by Daniel-Valentine — APPROVED** [review]

**Pull request merged**